## Day 46 Lecture 2 Assignment

In this assignment, we will calculate a similarity matrix using responses to a survey about student life at a university. We will perform clustering on this dataset later on.

In [1]:
%matplotlib inline

import matplotlib.pyplot as plt
import numpy as np
import pandas as pd
from scipy.spatial.distance import pdist, squareform

This dataset consists of 35 binary features, each corresponding to a yes/no question that characterizes the student taking the survey.

This dataset contains a large number of features, each corresponding to a survey question. The feature name summarizes the survey question, so we will not list them all out here.

Load the dataset.

In [2]:
# answer goes here
df = pd.read_csv('https://tf-assets-prod.s3.amazonaws.com/tf-curric/data-science/Data%20Sets%20Clustering/student_life_survey.csv')

df.info()
df.head()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2958 entries, 0 to 2957
Data columns (total 36 columns):
 #   Column                                                                                                                      Non-Null Count  Dtype
---  ------                                                                                                                      --------------  -----
 0   Q2-Participated in Societies and Interest Groups                                                                            2958 non-null   int64
 1   Q2-Participated in Clubs                                                                                                    2958 non-null   int64
 2   Q2-Participated in Halls, JCRCs and/or Residential College CSCs                                                             2958 non-null   int64
 3   Q2-Participated in University organised events                                                                              2958 non-

Q2-Participated in Societies and Interest Groups  ...  response_id
0                                                 0  ...            1
1                                                 0  ...            2
2                                                 0  ...            3
3                                                 1  ...            4
4                                                 1  ...            5

[5 rows x 36 columns]

For our analysis, we will focus on a specific subset of the survey that is focused on stress. These questions all begin with the string 'Q5'. Filter the columns that meet this criteria (should be 10 in total).

In [9]:
# answer goes here
stress = df.iloc[: , 25:35].copy()
stress


Q5-Stressed about Adjustment issues  ...  Q5-Stressed about Others
0                                       0  ...                         0
1                                       0  ...                         0
2                                       0  ...                         0
3                                       1  ...                         0
4                                       0  ...                         0
...                                   ...  ...                       ...
2953                                    0  ...                         0
2954                                    0  ...                         0
2955                                    0  ...                         0
2956                                    0  ...                         0
2957                                    0  ...                         0

[2958 rows x 10 columns]

We'll assume for the next step that a pair of negative values (i.e. both responses are "no") is as informative as a pair of positive values. Compute the full distance/dissimilarity matrix for the survey data using matching/hamming distance.

The pdist function is a simple way to accomplish this goal, and the squareform function will transform the output of the pdist function into a tabular format that can be convert to a dataframe more easily.

In [22]:
# answer goes here
matrix = pd.DataFrame(squareform(pdist(stress, metric='hamming')),index=stress.index, columns=stress.index)

np.fill_diagonal(matrix.values, np.nan)
matrix



0     1     2     3     4     5     ...  2952  2953  2954  2955  2956  2957
0      NaN   0.0   0.1   0.6   0.3   0.1  ...   0.3   0.2   0.0   0.0   0.2   0.2
1      0.0   NaN   0.1   0.6   0.3   0.1  ...   0.3   0.2   0.0   0.0   0.2   0.2
2      0.1   0.1   NaN   0.7   0.4   0.2  ...   0.4   0.3   0.1   0.1   0.1   0.1
3      0.6   0.6   0.7   NaN   0.5   0.5  ...   0.3   0.6   0.6   0.6   0.6   0.8
4      0.3   0.3   0.4   0.5   NaN   0.2  ...   0.2   0.3   0.3   0.3   0.5   0.3
...    ...   ...   ...   ...   ...   ...  ...   ...   ...   ...   ...   ...   ...
2953   0.2   0.2   0.3   0.6   0.3   0.3  ...   0.5   NaN   0.2   0.2   0.4   0.2
2954   0.0   0.0   0.1   0.6   0.3   0.1  ...   0.3   0.2   NaN   0.0   0.2   0.2
2955   0.0   0.0   0.1   0.6   0.3   0.1  ...   0.3   0.2   0.0   NaN   0.2   0.2
2956   0.2   0.2   0.1   0.6   0.5   0.3  ...   0.5   0.4   0.2   0.2   NaN   0.2
2957   0.2   0.2   0.1   0.8   0.3   0.3  ...   0.5   0.2   0.2   0.2   0.2   NaN

[2958 rows x 2958 columns]

Print the answers to the following three questions:

1. How many pairs of students are there in our dataset? (Remember to ignore a student pairing with themselves, and avoid double-counting - '3-5' is the same as '5-3'.) 
2. Of these pairs, how many gave identical answers? 
3. How many pairs had no answers in common?

Tip: it may help to first convert the NxN similarity matrix into a table with three columns: response_id_1, response_id_2, and the similarity between them. The new table should have NxN rows and 3 columns.

In [26]:
# combos of unqiue pairs = n(n-1)/2
matrix.shape[0]*(matrix.shape[0]-1)/2

4373403.0

In [64]:
count = matrix.apply(pd.value_counts).T
count

0.0    0.1    0.2  ...    0.8   0.9  1.0
0     341.0  717.0  804.0  ...   21.0   NaN  NaN
1     341.0  717.0  804.0  ...   21.0   NaN  NaN
2     260.0  696.0  763.0  ...    1.0   NaN  NaN
3       1.0   16.0   68.0  ...  202.0  11.0  NaN
4      12.0  105.0  491.0  ...    NaN   NaN  NaN
...     ...    ...    ...  ...    ...   ...  ...
2954  341.0  717.0  804.0  ...   21.0   NaN  NaN
2955  341.0  717.0  804.0  ...   21.0   NaN  NaN
2956   18.0  334.0  775.0  ...    2.0   NaN  NaN
2957  124.0  575.0  891.0  ...    2.0   NaN  NaN
avg     NaN    NaN    NaN  ...    NaN   NaN  NaN

[2959 rows x 326 columns]

In [66]:
# identical answers
count[0.0].sum()/2

136732.0

In [67]:
# How many pairs had no answers in common
count[1.0].sum()/2

291.0

For each student, calculate their average similarity with all other students. Then print the following:

1. On average, which student (i.e. which response_id) had the most similar responses to the survey population as a whole? What was the average similarity for this student, and what were their answers?
2. Repeat part 1, but for the student with the least similar responses to the survey population as a whole.
3. What do we notice about these students? What broader inferences can we make about the student population as a whole?

In [42]:
# answer goes here
matrix['avg'] = matrix.mean(axis=1)
matrix.avg.sort_values()



0       0.224822
440     0.224822
1865    0.224822
1869    0.224822
433     0.224822
          ...   
866     0.593777
878     0.593777
1796    0.593777
2416    0.601353
2899    0.624146
Name: avg, Length: 2958, dtype: float64

In [56]:
matrix.avg.value_counts()

0.224822    342
0.233818    261
0.247616    142
0.256611    125
0.263510    107
           ... 
0.400947      1
0.412851      1
0.501048      1
0.456409      1
0.516808      1
Name: avg, Length: 315, dtype: int64

In [51]:
most_similiar = matrix.avg.min()
most_similiar

0.2248224551910744

In [58]:
most_similiar_scores = stress.loc[:0]
most_similiar_scores

Q5-Stressed about Adjustment issues  ...  Q5-Stressed about Others
0                                    0  ...                         0

[1 rows x 10 columns]

342 students had the most similiar responses. The average similiarity was .2248. These students were most stressed about academic issues.

In [60]:
matrix.avg.sort_values(ascending=False).head()

2899    0.624146
2416    0.601353
1796    0.593777
878     0.593777
866     0.593777
Name: avg, dtype: float64

In [54]:
least_similiar = matrix.avg.max()
least_similiar

0.6241460940142

In [61]:
least_similiar_scores = stress.loc[2899:2899]
least_similiar_scores

Q5-Stressed about Adjustment issues  ...  Q5-Stressed about Others
2899                                    1  ...                         1

[1 rows x 10 columns]

1 student (id 2899) was the least similair of all the other students. That student has a avg sim score of .6241. They were stressed in 9/10 categories. The only category they weren't stressed in was romantic relationships. 

In [46]:
matrix.avg.mean()

0.31433517560581314

In [62]:
matrix.avg.value_counts().head(20)

0.224822    342
0.233818    261
0.247616    142
0.256611    125
0.263510    107
0.270003     86
0.272506     58
0.278999     55
0.279878     49
0.270883     45
0.286304     44
0.309571     42
0.301792     36
0.295299     34
0.292797     34
0.327156     30
0.308691     29
0.347988     29
0.291715     29
0.332093     29
Name: avg, dtype: int64

The average mean sim score was .3143 on a scale of 0 to 1. On the whole the students skew towards similiar answers than dissimilar ones.